File: param_management_ros2_integration_cpp/examples/default.cpp
// Copyright 2024 Simon Sagmeister
#include <iostream>
#include <param_management_cpp/base.hpp>
#include <param_management_cpp/param_value_manager.hpp>
#include <param_management_ros2_integration_cpp/helper_functions.hpp>
#include <rclcpp/node.hpp>
#include <rclcpp/rclcpp.hpp>
class MySoftwareModule
{
private:
  tam::pmg::ParamValueManager::SharedPtr param_manager_ =
    std::make_shared<tam::pmg::ParamValueManager>();
  double val_;
  void update_param_value()
  {
    val_ = param_manager_
             ->declare_and_get_value("a", 3.14, tam::pmg::ParameterType::DOUBLE, "Testparameter")
             .as_double();
  }

public:
  MySoftwareModule()
  {
    // CAREFUL: Independent from the overwrite, you will always get the default value here
    // since you create your module before connecting with a node or loading parameters
    update_param_value();
    std::cout << "Value (Constructor): " << val_ << std::endl;
  }
  void step()
  {
    update_param_value();
    std::cout << "Value (Step): " << val_ << std::endl;
  }
  tam::pmg::MgmtInterface::SharedPtr get_param_manager() const { return param_manager_; }
};
class ExampleNode : public rclcpp::Node
{
private:
  MySoftwareModule mod_;

public:
  explicit ExampleNode(MySoftwareModule const && module)
  : rclcpp::Node("TestNode"), mod_{std::move(module)}
  {
    // Connect your param manager to the callback
    // Important: don't forget to store your callback handle here, otherwise it will get deallocated
    auto callback_handle =
      tam::pmg::connect_param_manager_to_ros_cb(this, mod_.get_param_manager());
    // Declare all paramters from the param manager
    tam::pmg::declare_ros_params_from_param_manager(this, mod_.get_param_manager().get());

    // From now on, your module is supplied with the correct parameters from the overwrite file
    mod_.step();
  }
};
int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<ExampleNode>(MySoftwareModule());
  rclcpp::spin(node);
  rclcpp::shutdown();
}

File: param_management_ros2_integration_py/examples/default.py
# Copyright 2024 Simon Sagmeister
import rclpy
from rclpy.node import Node
from param_management_ros2_integration_py.helper_functions import (
    connect_param_manager_to_ros_cb,
    declare_ros_params_from_param_manager,
)
import param_management_py as pmg


class ROSIndependentSoftwareModule:

    def __init__(self):

        # This is an exemplary parameter that is only evaluated during creation of the node
        self.param_manager = pmg.ParamValueManager()
        self.value = None

        self._update_params()
        print("Value (Constructor): ", self.value)

        # Declare your parameters
        # CAREFUL: This will always return the default value here, since you
        # create the object before connecting it with the ros2 node

    def _update_params(self):
        self.value = self.param_manager.declare_and_get_value(
            "a", 3.14, pmg.ParameterType.DOUBLE, "A double parameter"
        ).as_double()

    def step(self):
        self._update_params()
        print("Value (Step): ", self.value)

    def get_param_manager(self):
        return self.param_manager


class ExampleNode(Node):

    def __init__(self, module: ROSIndependentSoftwareModule):
        super().__init__("TestNode")

        self.module = module

        connect_param_manager_to_ros_cb(
            self, self.module.get_param_manager()
        )  # Connect the param manager with the node -> Live updates
        declare_ros_params_from_param_manager(
            self, self.module.get_param_manager()
        )  # Declare all parameters from the param manager

        # From this point on, the module is parametrized correctly and is kept in sync
        self.module.step()


if __name__ == "__main__":
    rclpy.init()
    mod = ROSIndependentSoftwareModule()
    node = ExampleNode(mod)
    rclpy.spin(node)
    rclpy.shutdown()
